### ISSUE TITLE
Storybook 7 Support?

### ISSUE THREAD

**Issue by Opti-Digital** [UX-and-I/storybook-design-token#157]
I've been attempting to use the addon with storybook 7 but am getting errors when trying to setup the DocsPage for the design tokens using <DesignTokenDocBlock />. I've also needed to override the requirements for the @storybook/addons + addon-docs + components + theming to get rid of some errors with dependencies. 

However now I am getting an error in console on the .mdx DocsPage:
```ReactDOM.render is no longer supported in React 18. Use createRoot instead.```

Also receiving this error in console:
```Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component.```

Components addon section works but there is also a warning as below in console:
```./node_modules/@storybook/addon-docs/dist/chunk-R4NKYYJA.mjs 1:37-44Critical dependency: require function is used in a way in which dependencies cannot be statically extracted./node_modules/@storybook/addon-docs/dist/chunk-R4NKYYJA.mjs 1:166-173Critical dependency: require function is used in a way in which dependencies cannot be statically extracted```

Is there any plans to support Storybook 7 before it comes out off Beta? (Which I believe it is planned to be released in March)

**Comment by v11ncent** [comment]
Using `create-react-app` & Storybook 7, it added a whopping 12 high severity security vulnerabilities.

**Comment by Sqrrl** [comment]
Will look into this. Can't say for sure I'll get to it before 7 leaves beta, but I'll try.

**Comment by tupkovic** [comment]
> Using `create-react-app` & Storybook 7, it added a whopping 12 high severity security vulnerabilities.

I didn't use create-react-app, but rather just built using react and react-dom as dependencies (I am building a component library). I had 3 high vulnerabilities, however it is still in Beta officially so I assume some fixes will occur for this before official release anyway.

Sqrrl - Thanks! :) Keep us updated how you go, or if any help is needed

**Comment by satishvattikuti** [comment]
@Sqrrl does this add-on work out of the box with storybook-7? or is it still being developed?

**Comment by Sqrrl** [comment]
> @Sqrrl does this add-on work out of the box with storybook-7? or is it still being developed?

No, the addon won't work with SB 7 as is. I'm on it.

**Comment by satishvattikuti** [comment]
Thank you for the update @Sqrrl. Best of Luck

**Comment by Integrayshaun** [comment]
Hey @Sqrrl 👋 

I'm a Storybook DX engineer 😁 Let me know if you need any support in getting your addon upgraded to 7.0

**Comment by Sqrrl** [comment]
Thanks for reaching out. I'm a little behind schedule, but working on it. I've just published `v3.0.0-beta.0` with Storybook 7 support. 

For anyone interested, you can check out the v3 branch and install it with `npm install --save-dev storybook-design-token@next` or `yarn add storybook-design-token@next`.

**Comment by Sqrrl** [comment]
> Just installed this with `@storybook/angular` and got these missing dependency errors:
> 
> ```shell
> ERR! Error: Build failed with 3 errors:
> ERR! node_modules/storybook-design-token/dist/components/ClipboardButton.js:2:25: ERROR: Could not resolve "react-use-clipboard"
> ERR! node_modules/storybook-design-token/dist/components/TokenTable.js:4:27: ERROR: Could not resolve "react-virtual"
> ERR! node_modules/storybook-design-token/dist/hooks/useTokenTabs.js:2:32: ERROR: Could not resolve "react-storage-hooks"
> ```
> 
> Installing these manually worked, and now my doc blocks are working again. Thanks!

That's weird. The dependencies are in the package.json. Have you tried a clean install of your node_modules?

**Comment by dmytroyaremenkoleanix** [comment]
Tried 3.0.0-beta.0 with storybook 7 with Angular, have two issues:
- design-tokens.source.json wasn't generated (not for build, not for dev modes)
- when I use build command it shows error that vite module not found (but build is successful despite of that)

**Comment by dmytroyaremenkoleanix** [comment]
@Sqrrl if I edit `node_modules/storybook-design-token/dist/preset.js` and remove vite dependency usage , it works

**Comment by nategreen-dh** [comment]
@Sqrrl Looks like [they're in `devDependencies`](https://github.com/UX-and-I/storybook-design-token/blob/835d90e0e6ad67124da3b187653e56646ac857d7/addon/package.json#L57-L59) but since they're needed at runtime (for downstream projects) they need to be moved to either `dependencies` or `peerDependencies`. I'm guessing only those three showed up because the other React-oriented packages are already required by Storybook or something.

**Comment by nategreen-dh** [comment]
Actually, I think I had also already installed `vite` (as @dmytroyaremenkoleanix mentioned above), `react`, `react-dom` --- possibly for this addon @2.x, but maybe for a different addon (maybe it was `@storybook-extras/preset`...I don't recall).

**Comment by Sqrrl** [comment]
You can try `3.0.0-beta.2`. I got a little frustrated with the setup and switched to using the SB addon kit. I hope this resolves some issues.

Breaking change: Doc blocks have to be imported via `import { DesignTokenDocBlock } from 'storybook-design-token';` instead of `import { DesignTokenDocBlock } from 'storybook-design-token/dist/doc-blocks';`

**Comment by julrich** [comment]
I got it working with `3.0.0-beta.2` and `@storybook/react-vite` as the `framework`.

I had to manually fix a packaging problem (I think?) with the written `exports`. Vite was throwing an error because of this line:
https://github.com/UX-and-I/storybook-design-token/blob/v3/addon/package.json#L38

At first I thought that was only a typo... but changing this to `"./doc-blocks": {` did not help, either. When changing it to `"./dist/doc-blocks": {` everything works! Pretty sure that's just fixing symptoms, though. Only related thing I could find was:
https://github.com/UX-and-I/storybook-design-token/blob/v3/addon/src/index.ts#L1

... where the export doesn't specify a file extension. I think that could be a problem for Vite?

It gets written as `"export { DesignTokenDocBlock } from \"./doc-blocks\";` in the installed / compiled package, as part of the `index.js.map`.

The error when not fixing this looks like this with Vite:
![image](https://user-images.githubusercontent.com/136760/232238939-d3fbf205-5c0b-4447-9538-a35be15a1767.png)

**Comment by dmytroyaremenkoleanix** [comment]
Another question about 3.0.0-beta.2, it contains path-scurry dependency, which has BlueOak-1.0.0 license... Any idea whats this? Is this still keeps stroybook-design-token as MIT ?

**Comment by nategreen-dh** [comment]
I got `3.0.0-beta.2` working as well, removed my `vite` dev dependency but had to add all of these, which seems like it may have been overkill (not sure, so leaving it up to you):

```json
  "peerDependencies": {
    "@storybook/blocks": "^7.0.0",
    "@storybook/components": "^7.0.0",
    "@storybook/core-events": "^7.0.0",
    "@storybook/manager-api": "^7.0.0",
    "@storybook/preview-api": "^7.0.0",
    "@storybook/theming": "^7.0.0",
    "@storybook/types": "^7.0.0",
    "react": "^16.8.0 || ^17.0.0 || ^18.0.0",
    "react-dom": "^16.8.0 || ^17.0.0 || ^18.0.0"
  },
```

I noticed that the tabs are not working as of 3.x thus far, in case you hadn't seen this yet:

<img width="997" alt="image" src="https://user-images.githubusercontent.com/110842136/232501486-1a16e496-75d9-4deb-9235-b119ff9273cb.png">

I could write that up as a separate issue if that would help you. Thanks for your work on this plugin @Sqrrl , it's by far the best one out there right now for this purpose!

**Comment by aimee-gm** [comment]
`3.0.0-beta.2` is currently broken for me due to a typo in the `exports`.

The typo is `doc-blocks` is incorrectly `doc-bocks`.

In the published beta version, both the name of the export & the resolves files are incorrectly spelt.

In the branch, the name of the export is incorrect, and the corrected filenames are not published to a beta version.

https://github.com/UX-and-I/storybook-design-token/blob/v3/addon/package.json#L38

**Comment by Sqrrl** [comment]
Fixed with `3.0.0-beta.3`. Thanks for the PR, @bmalinconico.

**Comment by Jeremy-Walton** [comment]
I'm still seeing this error while using `3.0.0-beta.3` against Storybook `7.0.9`

`Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component.`

I've tried both `import { DesignTokenDocBlock } from 'storybook-design-token';` and `import { DesignTokenDocBlock } from 'storybook-design-token/doc-blocks';` but get the same issue. The tokens panel and tabs work fine, it's just the doc block that causes the error.

**Comment by Sqrrl** [comment]
Could you provide some more hints for reproduction? Storybook allows for so many different setups and frameworks, making it hard to reproduce such issues.

**Comment by Jeremy-Walton** [comment]
I was able to figure it out. There was a dependency resolution conflict which meant it didn't update the package correctly.
Clearing my lock file of the old version and readding it fixed the issues.

**Comment by tupkovic** [comment]
@Sqrrl - Managed to everything work nice and easy with beta.3 :) I have one issue though, it seems the disable of the tab is not working with Storybook 7 i guess. I have disabled as per documentation I think, but the tab does not get removed (it is not the default tab that shows though). Any ideas on this one?

I am using code below in preview.js to disable the tab as we only need the docs side of things because we have a theme switcher within stories and require style guide docs instead (via cssVariablesTheme).

`export const parameters = {
  designToken: {
    disable: true, /* Disabled as we will only use Doc Blocks for Documentation Pages */
  }
};`

**Comment by tupkovic** [comment]
> @Sqrrl - Managed to everything work nice and easy with beta.3 :) I have one issue though, it seems the disable of the tab is not working with Storybook 7 i guess. I have disabled as per documentation I think, but the tab does not get removed (it is not the default tab that shows though). Any ideas on this one?
>

Anyone else experienced this? Able to provide a work around? I have not been able to get around this disable tab issue yet so have had to remove the plugin from use unfortunately.

**Comment by Opti-Digital** [comment]
@Sqrrl Hi, any chance that the tab issue has been looked into or will be? I am also experiencing the same.

**Comment by allypalanzi** [comment]
Anyone have a working example of the beta for storybook 7 working with vite? i can't get my tokens to show up.

preview.ts
```
import type { Preview } from "@storybook/react";
import "../src/style.css";

const preview: Preview = {
  parameters: {
    actions: { argTypesRegex: "^on[A-Z].*" },
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/,
      },
    },
  },
};

export default preview;
```

main.ts
```
import type { StorybookConfig } from "@storybook/react-vite";

const config: StorybookConfig = {
  stories: ["../src/**/*.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx|mdx)"],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
    { name: "storybook-design-token", options: { preserveCSSVars: true } },
  ],
  framework: {
    name: "@storybook/react-vite",
    options: {},
  },
  docs: {
    autodocs: "tag",
  },
};
export default config;

```

And my CSS Variables are imported into styles via an npm package, but has the comments in there: 
```
:where(html) {
  /**
  * @tokens Brand Colors
  * @presenter Color
  */
  --pink-6: #67004c;

```

**Comment by dwightjack** [comment]
> > @Sqrrl - Managed to everything work nice and easy with beta.3 :) I have one issue though, it seems the disable of the tab is not working with Storybook 7 i guess. I have disabled as per documentation I think, but the tab does not get removed (it is not the default tab that shows though). Any ideas on this one?
> 
> Anyone else experienced this? Able to provide a work around? I have not been able to get around this disable tab issue yet so have had to remove the plugin from use unfortunately.

@tupkovic (cc @Sqrrl ). I have the same issue with SB7 and vite. I submitted a PR that should fix the issue: https://github.com/UX-and-I/storybook-design-token/pull/180

**Comment by Sqrrl** [comment]
> > > @Sqrrl - Managed to everything work nice and easy with beta.3 :) I have one issue though, it seems the disable of the tab is not working with Storybook 7 i guess. I have disabled as per documentation I think, but the tab does not get removed (it is not the default tab that shows though). Any ideas on this one?
> > 
> > 
> > Anyone else experienced this? Able to provide a work around? I have not been able to get around this disable tab issue yet so have had to remove the plugin from use unfortunately.
> 
> @tupkovic (cc @Sqrrl ). I have the same issue with SB7 and vite. I submitted a PR that should fix the issue: #180

Thank you. Merged and released with `3.0.0-beta.4`.

**Comment by Sqrrl** [comment]
I will close this issue now. Please open a new one, if you experience issues with Storybook 7.
